### DISCUSSION TITLE
Is there a way to pass arguments to `:Fnlfile`?

### DISCUSSION THREAD

**Discussion by datwaft** [rktjmp/hotpot.nvim#99] [Q&A]
Hello!

I am currently doing an small experiment, which is a template for plugins using Hotpot, but I am finding an small problem (You can find the WIP template here: [`datwaft/nvim-hotpot-plugin-template`](https://github.com/datwaft/nvim-hotpot-plugin-template)).

In my old plugin [`datwaft/themis.nvim`](https://github.com/datwaft/themis.nvim) I didn't use Hotpot for Unit Testing and so I didn't have access to Neovim's environment but with the experiment I am doing I am trying to get access to that environment using Hotpot.

My vision is, in the future, being able to use other, more powerful, testing frameworks but I am starting with [`andreyorst/fennel-test`](https://github.com/andreyorst/fennel-test).

When using [`andreyorst/fennel-test`](https://github.com/andreyorst/fennel-test) you usually execute a test like:

```sh
fennel \
  --add-fennel-path 'fnl/?.fnl' \
  --add-macro-path 'fnl/?.fnl' \
  --add-fennel-path 'deps/?.fnl' \
  --add-macro-path 'deps/?/init-macros.fnl' \
  deps/fennel-test/runner tests/hello-world.spec.fnl
```

I tried to do something similar with Hotpot and `:Fnlfile` like this:

```sh
nvim --headless \
  +"Fnlfile deps/fennel-test/runner tests/hello-world.spec.fnl" \
  +qa
```

But its displaying the following error message:

```traceback
Error detected while processing command line:
E5108: Error executing lua [string ":lua"]:1: runtime error: deps/fennel-test/runner tests/hello-world.spec.fnl: No such file or directory
stack traceback:
  [C]: in function 'assert'
  ....local/share/nvim/lazy/hotpot.nvim/lua/hotpot/fennel.lua:5139: in function ?
  [C]: in function 'eval_file'
  ...l/share/nvim/lazy/hotpot.nvim/lua/hotpot/api/command.lua:47: in function 'fnlfile'
  [string ":lua"]:1: in main chunk
stack traceback:
        [C]: in function 'fnlfile'
        [string ":lua"]:1: in main chunk
```

Do you have any idea about how to do this?

**PS:** and also any idea of how to use other testing frameworks with Hotpot?

**Comment by datwaft** [comment]
Seeing the latest breaking change, it seems that it works with `-l` for Lua files.

https://github.com/neovim/neovim/pull/18706#issue-1244364618

Maybe something similar can be done with Fennel files.

**Comment by rktjmp** [comment]
It all chains down to fennel dofile, so if you can make that accept args then hotpot can pass them through somehow, though it looks like it would be a breaking interface change unless the args were stuffed in the `options` argument.

Fennel may already pass the arguments, it should at least pass the modname and modfilepath.

Probably its simpler to do something like `:Fnl ((require :test-runner):run :test-name args)` or something and wrap that in a command.

**Reply by datwaft** [reply]
I thought about doing something like that but the runner automatically executes when requiring the module.

Investigating I found this article: https://enochchau.com/blog/2022/using-neovim-as-a-fennel-compiler/

In the article the author mentions a `:FnlRun` command, maybe we can add something like that to Hotpot.

```lua
command("FnlRun", function(t)
    local in_path = unpack(vim.fn.split(t.args, " "))
    assert(in_path, "missing input path")

    local out = fennel.dofile(in_path)
    -- print an extra newline to seperate script output from from the printed return values
    print("\n") 
    print_stdout(vim.inspect(out))
end, { nargs = 1 })
```

Reading https://fennel-lang.org/api#evaluate-a-file-of-fennel it seems you can pass arguments to `fennel.dofile`.

If you want I can try to create a PR for that, doesn't seem very difficult, just would need to know where to add new documentation.

**Comment by rktjmp** [comment]
Discussion not relevant to v2.
